File: java/src/com/nrkei/microservices/rapids_rivers/River.java
package com.nrkei.microservices.rapids_rivers;
/*
 * Copyright (c) 2016 by Fred George
 * May be used freely except for training; license required for training.
 * @author Fred George
 */

import com.google.gson.Gson;

import java.util.ArrayList;
import java.util.List;

// Understands a stream of valid JSON packets meeting certain criteria
// Implements GOF Observer pattern
// Implements GOF Command pattern
public class River implements Rapids.MessageListener {

    private final Rapids rapids;
    private final List<PacketListener> listeners = new ArrayList<>();
    private final List<Validation> validations = new ArrayList<>();

    public River(Rapids rapids) {
        this.rapids = rapids;
        rapids.register(this);
    }

    public void register(PacketListener listener) {
        listeners.add(listener);
    }

    @Override
    public void message(Rapids sendPort, String message) {
        PacketProblems problems = new PacketProblems(message);
        PacketBuilder builder = new PacketBuilder(message, problems);
        for (Validation v : validations) v.validate(builder);
        if (problems.hasErrors())
            onError(sendPort, problems);
        else
            packet(sendPort, builder);
    }

    private void packet(Rapids sendPort, PacketBuilder builder) {
        for (PacketListener l : listeners) l.packet(sendPort, builder.result());
    }

    private void onError(Rapids sendPort, PacketProblems problems) {
        for (PacketListener l : listeners) l.onError(sendPort, problems);
    }

    public void require(String jsonKey) {
        validations.add(new RequireKey(jsonKey));
    }

    public interface PacketListener {
        void packet(Rapids rapids, Packet packet);
        void onError(Rapids rapids, PacketProblems problems);
    }

    private interface Validation {
        void validate(PacketBuilder builder);
    }

    private class RequireKey implements Validation {
        private final String[] requiredKeys;

        RequireKey(String... requiredKeys) {
            this.requiredKeys = requiredKeys;
        }
        @Override
        public void validate(PacketBuilder builder) {
            builder.require(requiredKeys);
        }
    }

}

File: csharp_v2_beta/src/MicroServiceWorkshop/RapidsRivers/River.cs
/*
 * Copyright (c) 2016 by Fred George
 * May be used freely except for training; license required for training.
 */

using System;
using System.Collections.Generic;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace MicroServiceWorkshop.RapidsRivers
{
    public class River : RapidsConnection.IMessageListener
    {
        private readonly List<IPacketListener> _listeners = new List<IPacketListener>();
        private readonly List<IValidation> _validations = new List<IValidation>();

        public River(RapidsConnection rapidsConnection)
        {
            rapidsConnection.Register(this);
        }

        public void Register(IPacketListener listener)
        {
            _listeners.Add(listener);
        }
        
        public void HandleMessage(RapidsConnection sendPort, string message)
        {
            PacketProblems problems = new PacketProblems(message);
            JObject jsonPacket = JsonPacket(message, problems);
            foreach (IValidation v in _validations)
            {
                if (problems.AreSevere()) break;
                v.Validate(jsonPacket, problems);
            }
            if (problems.HasErrors())
                OnError(sendPort, problems);
            else
                Packet(sendPort, jsonPacket, problems);
        }

        private JObject JsonPacket(string message, PacketProblems problems)
        {
            JObject result = null;
            try
            {
                result = JObject.Parse(message);
            }
            catch (JsonException)
            {
                problems.SevereError("Invalid JSON format per NewtonSoft JSON library");
            }
            catch (Exception e)
            {
                problems.SevereError("Unknown failure. HandleMessage is: " + e.Message);
            }
            return result;
        }

        private void OnError(RapidsConnection sendPort, PacketProblems errors)
        {
            foreach (IPacketListener l in _listeners) l.OnError(sendPort, errors);
        }

        private void Packet(RapidsConnection sendPort, JObject jsonPacket, PacketProblems warnings)
        {
            foreach (IPacketListener l in _listeners) l.Packet(sendPort, jsonPacket, warnings);
        }

        public River Require(params string[] jsonKeyStrings)
        {
            _validations.Add(new RequiredKeys(jsonKeyStrings));
            return this;
        }

        public River Forbid(params string[] jsonKeyStrings)
        {
            _validations.Add(new ForbiddenKeys(jsonKeyStrings));
            return this;
        }

        public interface IPacketListener
        {
            void Packet(RapidsConnection connection, JObject jsonPacket, PacketProblems warnings);
            void OnError(RapidsConnection connection, PacketProblems errors);
        }

        private interface IValidation
        {
            void Validate(JObject jsonPacket, PacketProblems problems);
        }

        private class RequiredKeys : IValidation
        {
            private readonly string[] _requiredKeys;

            internal RequiredKeys(string[] requiredKeys)
            {
                _requiredKeys = requiredKeys;
            }

            public void Validate(JObject jsonPacket, PacketProblems problems)
            {
                foreach (string key in _requiredKeys)
                    if (jsonPacket[key] == null)
                        problems.Error("Missing required key '" + key + "'");
            }
        }

        private class ForbiddenKeys : IValidation
        {
            private readonly string[] _forbiddenKeys;

            internal ForbiddenKeys(string[] forbiddenKeys)
            {
                _forbiddenKeys = forbiddenKeys;
            }

            public void Validate(JObject jsonPacket, PacketProblems problems)
            {
                foreach (string key in _forbiddenKeys)
                {
                    JToken token = jsonPacket[key];
                    // Tests as suggested by NewtonSoft recommendations
                    if ((token == null) ||
                        (token.Type == JTokenType.Array && !token.HasValues) ||
                        (token.Type == JTokenType.Object && !token.HasValues) ||
                        (token.Type == JTokenType.String && token.ToString() == String.Empty) ||
                        (token.Type == JTokenType.Null))
                        problems.Information("Forbidden key '" + key + "' does not exist");
                    else problems.Error("Forbidden key '" + key + "' actually exists");
                }
            }
        }
    }
}